truncate/table_schema.go

```diff
@@ -137,6 +131,130 @@ func fetchTableSchemas(ctx context.Context, client *spanner.Client, targetTables
+// If targetTables is not empty, it fetches only the specified tables.
+// If excludeTables is not empty, it excludes the specified tables.
+// TargetTables and excludeTables cannot be specified at the same time.
+func filterTableSchemas(tables []*tableSchema, targetTables, excludeTables []string) ([]*tableSchema, error) {
+	isExclude := len(excludeTables) > 0
+	isTarget := len(targetTables) > 0
+
+	switch {
+	case isTarget && isExclude:
+		return nil, errors.New("both targetTables and excludeTables cannot be specified at the same time")
+	case isTarget:
+		return targetFilterTableSchemas(tables, targetTables), nil
+	case isExclude:
+		return excludeFilterTableSchemas(tables, excludeTables), nil
+	default: // No target and exclude tables are specified.
+		return tables, nil
+	}
+}
+
+// targetFilterTableSchemas filters tables with given targetTables.
+// If targetTables is empty, it returns all tables.
+func targetFilterTableSchemas(tables []*tableSchema, targetTables []string) []*tableSchema {
+	if len(targetTables) == 0 {
+		return tables
+	}
+
+	isTarget := make(map[string]bool, len(tables))
+	for _, t := range targetTables {
+		isTarget[t] = true
+	}
+
+	// TODO: Add child tables that may be deleted in cascade (#18)
+
+	filtered := make([]*tableSchema, 0, len(tables))
+	for _, t := range tables {
+		if isTarget[t.tableName] {
+			filtered = append(filtered, t)
+		}
+	}
+
+	return filtered
+}
+
+// excludeFilterTableSchemas filters tables with given excludeTables.
+// If excludeTables is empty, it returns all tables.
+// When an exclude table is cascade deletable, its parent table is also excluded.
+func excludeFilterTableSchemas(tables []*tableSchema, excludeTableSchemas []string) []*tableSchema {
+	if len(excludeTableSchemas) == 0 {
+		return tables
+	}
+
+	isExclude := make(map[string]bool, len(tables))
+	for _, t := range excludeTableSchemas {
+		isExclude[t] = true
+	}
+
+	// Additionally exclude parent tables that may delete the exclude tables in cascade
+	lineages := constructTableLineages(tables)
+	for _, l := range lineages {
+		if isExclude[l.tableSchema.tableName] {
+			for _, a := range l.ancestors {
+				isExclude[a.tableName] = true
+			}
+		}
+	}
+
+	filtered := make([]*tableSchema, 0, len(tables))
+	for _, t := range tables {
+		if !isExclude[t.tableName] {
+			filtered = append(filtered, t)
+		}
+	}
+
+	return filtered
+}
+
+// constructTableLineages returns a list of interleave Lineages.
+// This function creates tableLineage for each of all given tableSchemas.
+func constructTableLineages(tables []*tableSchema) []*tableLineage {
+	tableMap := make(map[string]*tableSchema, len(tables))
+	for _, t := range tables {
+		tableMap[t.tableName] = t
+	}
+
+	parentRelation := make(map[string]*tableSchema, len(tables))
+	childRelation := make(map[string][]*tableSchema, len(tables))
+	for _, t := range tables {
+		if !t.isRoot() && t.isCascadeDeletable() {
```

Comment:
If we construct lineages for tables, it might not be appropriate to use `t.isCascadeDeletable()`. Whether or not it's cascading deleted, I think the child table should be semantically in the table lineages.